### PR TITLE
Add many tests and fix bugs to meet 90% coverage

### PR DESCRIPTION
**Pull request by wkCircle**
The PR is mainly related to #20 , but also fix some bugs by the way. 

The PR 
- fixes the bug of `CollecitonUtils.prepend` (add the list(), same as the append method)
- adds its test in test_crud
- adds `__str__` and `pass` to white list of coverage in `pyproject.toml`.
- adds the test for `setHead`
- adds corner case inside `test_api_foldLeft_with_short_stop` 
- adds the test for `foldRight` 
- adds the test for `foldRight` with its arg `short_stop_func` 
- adds corner case inside `test_gapi_comp`
- tests also `iter` and its for loop of a flist inside `test_feat_for_in`
- tests the whole NIl class via the func `test_gapi_nil`
- adds the test for `getOrThrow`

The final test results (in python3.8): 
```zsh
$ make test 
coverage run -m pytest tests
====================================================================================== test session starts =======================================================================================
platform win32 -- Python 3.8.1, pytest-7.4.0, pluggy-1.2.0
configfile: pyproject.toml
collected 50 items

tests\unit\test_flist.py .........................                                                                                                                                          [ 50%]
tests\unit\test_fp.py .........................                                                                                                                                             [100%]

======================================================================================= 50 passed in 0.11s ======================================================================================= 
coverage report
Name              Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------------
fpu\__init__.py       2      2      0      0     0%   1-2
fpu\errors.py         5      0      0      0   100%
fpu\flist.py        178      6     46      1    97%   6-8, 297, 370, 386
fpu\fp.py           258     28     44      4    88%   7-8, 184, 197, 217, 237, 243-246, 249, 260, 263-267, 284-286, 289, 324, 350, 353, 356, 359, 388, 400, 416
fpu\utils.py         13      0      6      0   100%
-------------------------------------------------------------
TOTAL               456     36     96      5    92%
```

There are still some missing lines in both files since I am not sure what's the best way to solve them (not familiar with the class/func enough). Postpone them.

### CODE REVIEW

File: tests/unit/test_fp.py [hunk 1]
```diff
@@ -148,16 +150,20 @@ def test_crud(self):
...
+        self.assertEqual(ulist[0], 12)
+
     def test_first(self):
         alist = CollectionUtils.fl(self.tlist)
         olist = CollectionUtils.first(alist, lambda e: e % 2 == 0)
         self.assertTrue(len(olist) == 1, 'Expect to have 1 element (Real={})'.format(len(olist)))
         self.assertTrue(olist[0] == 2, 'Expect to have 2 (Real={})'.format(olist))
 
-    def test_fold(self):
+    def test_foldLeft(self):
```
Comment (wkCircle) [inline tests/unit/test_fp.py:166]: I rename it to distinguish it from newly added foldRight test cases

File: tests/unit/test_fp.py [hunk 2]
```diff
@@ -168,10 +174,26 @@ def test_fold(self):
...
         order = CollectionUtils.l(OrderLine(toothPaste, 2),
                                   OrderLine(toothBrush, 3))
-        weight = CollectionUtils.foldLeft(order, 0.0, lambda i, e: i + e.getWeight())
-        price = CollectionUtils.foldLeft(order, 0.0, lambda i, e: i + e.getAmount())
-        self.assertEqual(price, 13.5, 'Expect 13.5 (Real={})'.format(weight))
-        self.assertEqual(weight, 1.9, 'Expect 1.9 (Real={})'.format(price))
+        weight = CollectionUtils.foldLeft(order, 0.26, lambda i, e: i + e.getWeight())
+        price = CollectionUtils.foldLeft(order, 1.15, lambda i, e: i + e.getAmount())
+        self.assertEqual(price, 13.5 + 1.15, 'Expect 13.5 (Real={})'.format(weight))
+        self.assertEqual(weight, 1.9 + 0.26, 'Expect 1.9 (Real={})'.format(price))
```
Comment (wkCircle) [inline tests/unit/test_fp.py:180]: make initial value more diverse to truly test both foldLeft and foldRight

File: tests/unit/test_fp.py [hunk 3]
```diff
@@ -212,11 +234,11 @@ def test_range(self):
             self.assertTrue(olist[i] == i)
 
 class TailCallTest(unittest.TestCase):
-    r'''
+    '''
     Test Cases of class TailCall
     '''
     def setUp(self):
-        self.tlist = [1, 2, 3, 4]
+        pass
```
Comment (wkCircle) [inline tests/unit/test_fp.py:241]: not used in the class

File: tests/unit/test_fp.py [hunk 4]
```diff
@@ -243,12 +265,12 @@ def test_fib(self):
...
 class OptionTest(unittest.TestCase):
-    r'''
+    '''
     Test Cases of class Option
     '''
+
     def setUp(self):
-        self.tlist = [1, 2, 3, 4]
-        self.df = lambda e: e
+        pass
```
Comment (wkCircle) [inline tests/unit/test_fp.py:273]: remove self.df and define df in local since it is used only once.

File: tests/unit/test_flist.py [hunk 5]
```diff
@@ -82,30 +87,61 @@ def test_api_foldLeft(self):
...
+    """Testing API Cons.foldLeft with arg short_stop_func."""
     alist = fl(1, 2, 3, 4, 5)
-
-    result = alist.foldLeft(
-        0, lambda a, b: a + b,
-        short_stop_func=lambda v: v > 3)
+    f = lambda a, b: a + b # noqa
+    # test corner case where the func stops at identity
+    result = alist.foldLeft(0, f, short_stop_func=lambda v: v > -1)
+    self.assertIsInstance(result, Nil)
```
Comment (johnklee) [inline tests/unit/test_flist.py:95]: Brilliant idea to add test case to cover corner case.  Two feedback for your reference. 
1. This is a good usage case in using [parameterized testing feature](https://docs.pytest.org/en/7.1.x/example/parametrize.html). 
2. Please keep the unit test case content to follow [AAA structure](https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80)

Below is sample code for reference:
```python
@pytest.mark.parametrize(
    "indicator,short_stop_func,expected",
    [
        pytest.param(
            0, lambda v: v > 3, 6, id="FAST",
        ),
        pytest.param(
            0, lambda v: v > -1, flist.nil, id="corner_case",
        ),
    ])
def test_api_fold_left_with_short_stop(
    indicator, short_stop_func, expected):
  # Arrange
  alist = fl(1, 2, 3, 4, 5)

  # Action
  result = alist.foldLeft(
      0, lambda a, b: a + b,
      short_stop_func=short_stop_func)

  # Assert
  # The folding operation will stop at element 4.
  # So we have folding result as 0 + 1 + 2 + 3 = 6
  assert result == expected
```

File: tests/unit/test_flist.py [hunk 6]
```diff
@@ -82,30 +87,61 @@ def test_api_foldLeft(self):
...
 
     # The folding operation will stop at element 4.
     # So we have folding result as 0 + 1 + 2 + 3 = 6
+    result = alist.foldLeft(0, f, short_stop_func=lambda v: v > 3)
     self.assertEqual(result, 6)
 
+  def test_api_foldReft(self):
```
Comment (johnklee) [inline tests/unit/test_flist.py:102]: According to the comment, maybe the name of unit test case should be `test_api_foldRight`?

File: tests/unit/test_flist.py [hunk 7]
```diff
@@ -82,30 +87,61 @@ def test_api_foldLeft(self):
...
+    """Testing API Cons.foldRight."""
+    alist = fl(1, 2, 3, 4, 5)
+    self.assertEqual('123450', alist.foldRight(0, lambda a, e: "{}{}".format(a, e)))
+    self.assertEqual(15, alist.foldRight(0, lambda a, e: a + e))
+    self.assertEqual(120.0, alist.foldLeft(1.0, lambda a, e: a * e))
```
Comment (johnklee) [inline tests/unit/test_flist.py:107]: You might want to place this line in unit test case `test_api_foldLeft`

File: tests/unit/test_flist.py [hunk 8]
```diff
@@ -82,30 +87,61 @@ def test_api_foldLeft(self):
...
+    alist = fl(1, 2, 3, 4, 5)
+    f = lambda a, b: a + b # noqa
+    # test corner case where the func stops at identity
+    result = alist.foldRight(0, f, short_stop_func=lambda v: v > -1)
+    self.assertIsInstance(result, Nil)
+
+    # The folding operation will stop at element 4.
+    # So we have folding result as 3 + 4 + 5 = 15
+    result = alist.foldRight(0, f, short_stop_func=lambda v: v > 3)
+    self.assertEqual(result, 15)
```
Comment (johnklee) [inline tests/unit/test_flist.py:120]: This is actually two unit test cases. You could leverage "[Parametrizing test](https://docs.pytest.org/en/7.1.x/example/parametrize.html)" feature to use same unit test case with different testing data to cover them together (or more).

File: tests/unit/test_flist.py [hunk 9]
```diff
@@ -82,30 +87,61 @@ def test_api_foldLeft(self):
...
   def test_gapi_fl(self):
     """Testing global API:fl to create object of List."""
-    alist = fl(1, 2, 3)
-    self.assertEqual('[1, 2, 3, NIL]', str(alist))
-    self.assertEqual(3, alist.size())
-    self.assertEqual(3, alist.length())
-    self.assertEqual(1, alist.head())
-    self.assertEqual('[2, 3, NIL]', str(alist.tail()))
+    # case input element (int/float/str) + list + Generator
+    def firstevens(n: int):
```
Comment (johnklee) [inline tests/unit/test_flist.py:125]: Nit: Maybe `even_number_generator(until: int)`

File: tests/unit/test_flist.py [hunk 10]
```diff
@@ -82,30 +87,61 @@ def test_api_foldLeft(self):
...
   def test_gapi_comp(self):
     """Testing global API:comp to generate composition from given list."""
+    # corner case
+    fpu_comp_list = comp([], random.randint(0, 10))
+    self.assertIsInstance(fpu_comp_list, Nil)
+
     # Composition for 2 element of [1, 2, 3] will be [(1, 2), (1, 3), (2, 3)]
     fpu_comp_list = comp([1, 2, 3], 2)
-
     self.assertEqual('[(1, 2), (1, 3), (2, 3), NIL]', str(fpu_comp_list))
```
Comment (johnklee) [inline tests/unit/test_flist.py:145]: Refer to [`Parametrizing tests`](https://docs.pytest.org/en/7.1.x/example/parametrize.html). Thanks for adding more unit test case to increase coverage. Good job.

File: fpu/fp.py [hunk 11]
```diff
@@ -440,7 +440,7 @@ def reverse(alist):
 
   @staticmethod
   def prepend(alist, e):
-    nlist = CollectionUtils.copy(alist)
+    nlist = list(CollectionUtils.copy(alist))
     nlist.insert(0, e)
     return nlist
```
Comment (johnklee) [inline fpu/fp.py:445]: We want to keep container as immutable to meet FP requirement. So possibly, we could do this way:
```python
nlist = CollectionUtils.copy(alist)   # tuple instance
return (e, ) + nlist
```

File: tests/unit/test_fp.py [hunk 12]
```diff
@@ -168,10 +174,26 @@ def test_fold(self):
...
+    def test_foldRight(self):
+        alist = CollectionUtils.l(*self.tlist)
+        rst1 = CollectionUtils.foldRight(alist, 0, lambda i, e: i + e)
+        rst2 = CollectionUtils.foldRight(alist, 0, lambda e, i: i + e)
+        self.assertEqual(rst1, 10)  # 0 + 1 + 2 + 3 + 4 = 10
+        self.assertEqual(rst2, 10)  # 0 + 4 + 3 + 2 + 1 = 10
+
+        toothPaste = Product('Tooth paste', 1.5, 0.5)
+        toothBrush = Product('Tooth brush', 3.5, 0.3)
+        order = CollectionUtils.l(OrderLine(toothPaste, 2),
```
Comment (johnklee) [inline tests/unit/test_fp.py:191]: Nit: `orders`

File: tests/unit/test_fp.py [hunk 13]
```diff
@@ -168,10 +174,26 @@ def test_fold(self):
...
+                                  OrderLine(toothBrush, 3))
+        weight = CollectionUtils.foldRight(order, 0.26, lambda e, i: i + e.getWeight())
+        price = CollectionUtils.foldRight(order, 1.15, lambda e, i: i + e.getAmount())
+        self.assertEqual(price, 13.5 + 1.15, 'Expect 13.5 (Real={})'.format(weight))
```
Comment (johnklee) [inline tests/unit/test_fp.py:195]: Expect `14.65` (13.5 + 1.15)?

File: tests/unit/test_fp.py [hunk 14]
```diff
@@ -168,10 +174,26 @@ def test_fold(self):
...
+        self.assertEqual(weight, 1.9 + 0.26, 'Expect 1.9 (Real={})'.format(price))
```
Comment (johnklee) [inline tests/unit/test_fp.py:196]: Expect `2.16`? (1.9 + 0.26)

**Review by johnklee — CHANGES_REQUESTED** [review]
Some comments may deserve your attention. Thanks for adding more unit test cases to increase the coverage and improvement of the code quality.

**Review by johnklee — APPROVED** [review]
LGTM

File: tests/unit/test_flist.py [hunk 15]
```diff
@@ -82,30 +87,60 @@ def test_api_foldLeft(self):
     self.assertEqual(120.0, alist.foldLeft(1.0, lambda a, e: a * e))
 
   def test_api_foldLeft_with_short_stop(self):
+    """Testing API Cons.foldLeft with arg short_stop_func."""
+    # TODO: fpu/57 - pytest parameterize not supported in unittest class
```
Comment (johnklee) [inline tests/unit/test_flist.py:91]: Nit: If it makes sense to you, let's use Google style to make todo marker. e.g.:
```
# TODO(#57): pytest parameterize not supported in unittest class
```
FYI
